pkg/kubelet/server.go

```diff
@@ -587,80 +595,160 @@ func (s *Server) getPortForward(request *restful.Request, response *restful.Resp
+}
+
+func portForwardStreamReceived(streams chan httpstream.Stream) func(httpstream.Stream) error {
+	return func(stream httpstream.Stream) error {
 		portString := stream.Headers().Get(api.PortHeader)
 		port, err := strconv.ParseUint(portString, 10, 16)
 		if err != nil {
-			return fmt.Errorf("Unable to parse '%s' as a port: %v", portString, err)
+			return fmt.Errorf("unable to parse %q as a port: %v", portString, err)
 		}
 		if port < 1 {
-			return fmt.Errorf("Port '%d' must be greater than 0", port)
+			return fmt.Errorf("port %d must be greater than 0", port)
 		}
-		streamChan <- stream
+		streams <- stream
 		return nil
-	})
+	}
+}
+
+func getPortForwardStreamConnectionID(stream httpstream.Stream) string {
+	connectionID := stream.Headers().Get(api.ConnectionIDHeader)
+	if len(connectionID) == 0 {
+		glog.V(5).Infof("Stream received without connection ID header")
+		// old client that isn't generating connectionID
+		streamID := stream.Identifier()
+
+		// When there aren't concurrent new forwarded connections, each
+		// connection will have a pair of streams (data, error), and the stream
+		// IDs will be consecutive odd numbers, e.g. 1 and 3 for the first
+		// connection. Convert the stream ID into a pseudo-connectionID by dividing
+		// it by 4 and taking the floor. For example, streams 1 and 3 should
+		// both resolve to connectionID=0, streams 5 and 7 should both resolve to
+		// connectionID=1, and so on.
+		//
+		// NOTE: this only works when there are not concurrent new streams
+		// from multiple forwarded connections; it's a best-effort attempt
+		// at supporting old clients that don't generate connection IDs.
+		// If there are concurrent new connections, it's possible that 1
+		// connection gets streams whose IDs are not consecutive (e.g. 5
+		// and 9 instead of 5 and 7). In this case, it's likely that
+		// the client will be unable to forward any more ports until
+		// the `kubectl port-forward` command is terminated and restarted.
+		connectionID = strconv.Itoa(int(streamID) / 4)
+		glog.V(5).Infof("Automatically assigning connection ID %d from stream ID %d", connectionID, streamID)
+	}
+	return connectionID
+}
+
+// ServePortForward handles a port forwarding request.
+func ServePortForward(w http.ResponseWriter, req *http.Request, portForwarder PortForwarder, podName string, uid types.UID, idleTimeout time.Duration) {
+	streamChan := make(chan httpstream.Stream, 1)
+
+	glog.V(5).Infof("Upgrading port forward response")
+	upgrader := spdy.NewResponseUpgrader()
+	conn := upgrader.UpgradeResponse(w, req, portForwardStreamReceived(streamChan))
 	if conn == nil {
 		return
 	}
 	defer conn.Close()
-	conn.SetIdleTimeout(s.host.StreamingConnectionIdleTimeout())
 
-	var dataStreamLock sync.Mutex
-	dataStreamChans := make(map[string]chan httpstream.Stream)
+	glog.V(5).Infof("Setting port forwarding streaming connection idle timeout to %v", idleTimeout)
+	conn.SetIdleTimeout(idleTimeout)
 
+	var streamLock sync.Mutex
```

Comment:
`streamLock := &sync.Mutex{}`, or the pass-by-value to `handlePortForwardStreams` won't actually be using the same lock